lib/core/pmds/net/tap/pmd_tap.c

```diff
@@ -0,0 +1,381 @@
+#include <net/ethernet.h>           // for ether_addr
+#include <sys/uio.h>
+#include <linux/if_tun.h>        // for if_nametoindex, IF_NAMESIZE
+#include <tun_alloc.h>
+
+#include "pmd_tap.h"
+
+struct tap_rx_q {
+    int fd;
+    struct iovec *rd;
+
+    struct pmd_lport *lport;
+    uint16_t lport_id;
+
+    uint64_t n_pkts;
+    uint64_t n_bytes;
+};
+
+struct tap_tx_q {
+    int fd;
+    struct iovec *rd;
+
+    uint64_t n_pkts;
+    uint64_t n_bytes;
+};
+
+struct pmd_lport {
+    uint16_t lport_id;
+    char if_name[IFNAMSIZ + 1];
+    pktmbuf_info_t *pi;
+    struct tap_info *ti;
+    struct ether_addr eth_addr;
+
+    struct tap_rx_q *rxq;
+    struct tap_tx_q *txq;
+};
+
+static uint16_t
+pmd_tuntap_rx(void *queue, pktmbuf_t **bufs, uint16_t nb_pkts, int tap_type __cne_unused)
+{
+    struct tap_rx_q *rxq = queue;
+    int n_rx_pkts        = 0;
+    int n_rx_bytes       = 0;
+    struct tun_pi pi;
+    uint16_t len;
+    int nb_cnt = 0, k;
+
+    if (!rxq || !bufs || nb_pkts == 0)
+        return 0;
+
+    if ((nb_cnt = pktmbuf_alloc_bulk(rxq->lport->pi, bufs, nb_pkts)) < 0)
+        return 0;
+
+    /* Read packets from the TAP socket & store in allocated mbuf */
+    for (int i = 0; i < nb_cnt; i++) {
+        pktmbuf_t *m = *bufs;
+        struct iovec iov[2];
+
+        k = 0;
+
+        iov[k].iov_base  = &pi;
+        iov[k++].iov_len = sizeof(struct tun_pi);
+        iov[k].iov_base  = pktmbuf_mtod(m, void *);
+        iov[k++].iov_len = pktmbuf_tailroom(m);
+
+        len = readv(rxq->fd, iov, k);
+        if (len < sizeof(struct tun_pi) || len == 0xFFFF)
+            break;
+
+        if (pi.flags & TUN_PKT_STRIP)
+            continue;
+        len -= sizeof(struct tun_pi);
+
+        bufs++;
+
+        pktmbuf_port(m)     = rxq->lport_id;
+        pktmbuf_data_len(m) = len;
+
+        n_rx_pkts++;
+        n_rx_bytes += len;
+    }
+
+    if (n_rx_pkts < nb_cnt)
+        pktmbuf_free_bulk(bufs, nb_cnt - n_rx_pkts);
+
+    rxq->n_pkts += n_rx_pkts;
+    rxq->n_bytes += n_rx_bytes;
+
+    return n_rx_pkts;
+}
+
+static uint16_t
+pmd_tap_rx(void *queue, pktmbuf_t **bufs, uint16_t nb_pkts)
+{
+    return pmd_tuntap_rx(queue, bufs, nb_pkts, IFF_TAP);
+}
+
+static uint16_t
+pmd_tun_rx(void *queue, pktmbuf_t **bufs, uint16_t nb_pkts)
+{
+    return pmd_tuntap_rx(queue, bufs, nb_pkts, IFF_TUN);
+}
+
+static uint16_t
+pmd_tuntap_tx(void *queue, pktmbuf_t **bufs, uint16_t nb_pkts, int tap_type)
+{
+    struct tap_tx_q *txq = queue;
+    uint64_t tx_pkts = 0, tx_bytes = 0;
+
+    if (!txq || !bufs)
+        return 0;
+
+    if (nb_pkts) {
+        struct tun_pi pi = {.flags = 0, .proto = 0};
+        struct iovec iov[2];
+        int k;
+
+        for (int i = 0; i < nb_pkts; i++) {
+            pktmbuf_t *m = bufs[i];
+            ssize_t len;
+
+            pi.flags = 0;
+            pi.proto = 0;
+            if (tap_type == IFF_TUN) {
+                char proto = *pktmbuf_mtod(m, char *);
+
+                if (proto == 0x40)
```

Comment:
Is the IP protocol version supposed to be set in teh packet data? Is that what you're doing here? Shouldn't you mask off the version bits?

Reply:
Yes, should mask the bits for version.